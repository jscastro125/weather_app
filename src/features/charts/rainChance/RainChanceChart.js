import React from 'react';
import { useSelector } from 'react-redux';
import { ComposedChart, Area, XAxis, YAxis, Tooltip, CartesianGrid, Line, ResponsiveContainer } from 'recharts';

import {
    selectPressure
} from '../../pressure/pressureSlice';
import {
    selectTemperature
} from '../../temperature/temperatureSlice';
import {
    selectRainfallAmounts
} from '../rainfallAmount/rainfallSlice';
import {getDayName, chanceOfRain} from "../../../utils/utils";

export default function RainChanceChart() {
    const temperature = useSelector(selectTemperature);
    const pressure = useSelector(selectPressure);
    const rainfallAmounts = useSelector(selectRainfallAmounts);

    const data = rainfallAmounts.map(({ day, amount }) => {
        const [lower, mean, upper] = chanceOfRain(pressure, temperature, amount);
        return {
            day: getDayName(day),
            temperatureInterval: [lower, upper],
            temperature: mean
        }
    });
    return (
        <>
            <h3>
                Chance of rain
            </h3>
            <ResponsiveContainer width="100%" height={250}>
                <ComposedChart
                    data={data}
                >
                    <CartesianGrid stroke="#f5f5f5" />
                    <XAxis dataKey="day" />
                    <YAxis />
                    <Tooltip />
                    <Area type="monotone" dataKey="temperatureInterval" fill="#000" stroke="#000" />
                    <Line type="monotone" dataKey="temperature" stroke="#000" />
                </ComposedChart>
            </ResponsiveContainer>
        </>
    );
}
