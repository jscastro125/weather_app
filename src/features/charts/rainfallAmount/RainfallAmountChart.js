import React, {useEffect} from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {BarChart, Bar, XAxis, YAxis, CartesianGrid, Tooltip, ResponsiveContainer} from 'recharts';

import {
    fetchRainfallAmounts,
    selectRainfallAmounts
} from './rainfallSlice';
import {getDayName} from "../../../utils/utils";

export default function RainfallAmount() {
    const rainfallAmounts = useSelector(selectRainfallAmounts);
    const dispatch = useDispatch();
    useEffect(() => {
        dispatch(fetchRainfallAmounts());
    }, [dispatch]);
    const data = rainfallAmounts.map(({ day, amount }) => ({ day: getDayName(day), amount }));
    return (
        <>
            <h3>
                Amount of rainfall
            </h3>
            <ResponsiveContainer width="100%" height={250}>
                <BarChart data={data}>
                    <CartesianGrid strokeDasharray="3 3" />
                    <XAxis dataKey="day" />
                    <YAxis />
                    <Tooltip />
                    <Bar dataKey="amount" fill="#000" />
                </BarChart>
            </ResponsiveContainer>
        </>
    );
}
