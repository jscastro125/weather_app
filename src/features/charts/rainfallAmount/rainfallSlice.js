import { createSlice } from '@reduxjs/toolkit';

export const rainfallSlice = createSlice({
    name: 'rainfallAmounts',
    initialState: {
        value: [],
        isLoading: false,
        error: null
    },
    reducers: {
        setIsLoading: (state, action) => {
            state.isLoading = action.payload;
        },
        setError: (state, action) => {
            state.error = action.payload;
        },
        setValue: (state, action) => {
            state.value = action.payload;
        },
    },
});

export const { setIsLoading, setError, setValue } = rainfallSlice.actions;

export const fetchRainfallAmounts = () => dispatch => {
    dispatch(setIsLoading(true));
    fetch('https://private-4945e-weather34.apiary-proxy.com/weather34/rain')
        .then(res => res.json())
        .then(data => {
            if(data?.length !== 0) {
                const [currentWeek] = data;
                dispatch(setValue(currentWeek?.days));
                dispatch(setIsLoading(false));
            }
        })
        .catch(err => dispatch(setError(err)))
};

export const selectRainfallAmounts = state => state.rainfallAmounts.value;
export const selectError = state => state.rainfallAmounts.selectError;

export default rainfallSlice.reducer;
