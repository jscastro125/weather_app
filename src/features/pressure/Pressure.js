import React from 'react';
import { useSelector, useDispatch } from 'react-redux';

import {
    set,
    selectPressure,
} from './pressureSlice';
import Slider from "../../components/Slider/Slider";

export default function Pressure() {
    const pressure = useSelector(selectPressure);
    const dispatch = useDispatch();
    return (
        <Slider label="Pressure [ hPa ]" max={1030} min={970} value={pressure} onChange={(nPressure) => {
            dispatch(set(nPressure))
        }}/>
    );
}
