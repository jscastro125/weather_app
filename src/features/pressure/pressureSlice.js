import { createSlice } from '@reduxjs/toolkit';

export const pressureSlice = createSlice({
    name: 'pressure',
    initialState: {
        value: 1010,
    },
    reducers: {
        set: (state, action) => {
            state.value = action.payload;
        },
    },
});

export const { set } = pressureSlice.actions;

export const selectPressure = state => state.pressure.value;

export default pressureSlice.reducer;
