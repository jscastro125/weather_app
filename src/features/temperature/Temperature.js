import React from 'react';
import { useSelector, useDispatch } from 'react-redux';

import {
    set,
    selectTemperature,
} from './temperatureSlice';
import Slider from "../../components/Slider/Slider";

export default function Temperature() {
    const temperature = useSelector(selectTemperature);
    const dispatch = useDispatch();
    return (
        <Slider label="Temperature" max={35} min={10} value={temperature} onChange={(nTemperature) => {
            dispatch(set(nTemperature))
        }}/>
    );
}
