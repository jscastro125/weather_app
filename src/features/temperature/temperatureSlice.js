import { createSlice } from '@reduxjs/toolkit';

export const temperatureSlice = createSlice({
    name: 'temperature',
    initialState: {
        value: 15,
    },
    reducers: {
        set: (state, action) => {
            state.value = action.payload;
        },
    },
});

export const { set } = temperatureSlice.actions;

export const selectTemperature = state => state.temperature.value;

export default temperatureSlice.reducer;
