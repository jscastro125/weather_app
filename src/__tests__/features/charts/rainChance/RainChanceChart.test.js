import React from "react";
import { create } from "react-test-renderer";
import { useSelector } from 'react-redux';

import rainfallAmountsMock from '../../../../__mocks__/rainfallAmounts.mock';
import * as pressureSlice from '../../../../features/pressure/pressureSlice';
import RainChanceChart from '../../../../features/charts/rainChance/RainChanceChart';
import * as temperatureSlice from '../../../../features/temperature/temperatureSlice';
import * as amountsSlice from '../../../../features/charts/rainfallAmount/rainfallSlice';

jest.mock('react-redux', () => ({
    useSelector: jest.fn(),
}));

function createNodeMock() {
    const doc = document.implementation.createHTMLDocument();
    return { parentElement: doc.body };
}

describe("<RainChanceChart />", () => {
    it("should match snapshot", () => {
        useSelector.mockImplementation(sel => sel());
        const spyPressure = jest.spyOn(pressureSlice, 'selectPressure');
        spyPressure.mockReturnValue(30);
        const spyTemperature = jest.spyOn(temperatureSlice, 'selectTemperature');
        spyTemperature.mockReturnValue(1010);
        const spyAmounts = jest.spyOn(amountsSlice, 'selectRainfallAmounts');
        spyAmounts.mockReturnValue(rainfallAmountsMock);
        window.innerWidth = 1100;
        const component = create(<RainChanceChart />, { createNodeMock });
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
});
