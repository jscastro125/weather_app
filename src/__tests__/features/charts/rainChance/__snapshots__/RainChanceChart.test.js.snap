// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RainChanceChart /> should match snapshot 1`] = `
Array [
  <h3>
    Chance of rain
  </h3>,
  <div
    className="recharts-responsive-container"
    style={
      Object {
        "height": 250,
        "maxHeight": undefined,
        "minHeight": undefined,
        "minWidth": undefined,
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "display": "none",
          "height": 0,
          "position": "absolute",
          "visibility": "hidden",
          "width": 0,
        }
      }
    />
  </div>,
]
`;
